### PR TITLE
Fixes for model import, test refactoring

### PR DESCRIPTION
**Pull request by agibsonccc**
## What changes were proposed in this pull request?
1. Changes around TF import including:
   allow empty arrays as outputs and add more flexibility around shape validation
2. Minor fix around dimension computation: one edge case occurs in reduction shape calculation at java level when -1 is specified for dimension
3. Add select as control flow omission for eager mode
4. Disables tests that take download files: may refactor/group tests to allow selective running. Download tests need to be revisited.
5. Back track line search bug fix: converged early returning wrong results.
6. Add more flexibility around broadcasting at c++ level for ops like assign.

(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

## Quick checklist

The following checklist helps ensure your PR is complete:

- [ ] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [ ] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [ ] Created tests for any significant new code additions.
- [ ] Relevant tests for your changes are passing.

### CODE REVIEW

File: deeplearning4j/deeplearning4j-data/deeplearning4j-datasets/src/main/java/org/deeplearning4j/datasets/fetchers/MnistDataFetcher.java [hunk 1]
```diff
@@ -80,42 +81,52 @@ public MnistDataFetcher(boolean binarize) throws IOException {
...
-            } catch (Exception e2){ }
-            new MnistFetcher().downloadAndUntar();
-            MnistManager man = new MnistManager(images, labels, train);
-            lastCursor = man.getCurrent();
-            validateFiles(files, checksums);
-            man.close();
+            e.printStackTrace();
         }
 
         MnistManager man = new MnistManager(images, labels, train);
```
Comment (treo) [inline deeplearning4j/deeplearning4j-data/deeplearning4j-datasets/src/main/java/org/deeplearning4j/datasets/fetchers/MnistDataFetcher.java]: Why still **two** MnistManager instantiations?

File: libnd4j/include/ops/declarable/generic/broadcastable/floormod.cpp [hunk 2]
```diff
@@ -66,19 +66,20 @@ CUSTOM_OP_IMPL(floormod_bp, 3, 2, false, 0, 0) {
...
+    epsNext->applyPairwiseTransform(pairwise::Multiply, temp2, *gradY);
   else  // epsNext is greater than gradY
   {
     std::vector<sd::LongType> dims(epsNext->rankOf() * 2);
     sd::LongType gap = epsNext->rankOf() - gradY->rankOf();
     for (sd::LongType d = 0; d < gap; d++) {
       dims[d * 2 + 1] = 1;
     }
-    auto tempIn((temp)(dims));
+    auto tempIn((temp2)(dims));
```
Comment (treo) [inline libnd4j/include/ops/declarable/generic/broadcastable/floormod.cpp]: Any reason to rename the variable here?

File: platform-tests/src/test/java/org/eclipse/deeplearning4j/dl4jcore/datasets/iterator/DataSetIteratorTest.java [hunk 3]
```diff
@@ -66,9 +69,12 @@
 @Tag(TagNames.LONG_TEST)
 class DataSetIteratorTest extends BaseDL4JTest {
 
+
+
+
     @Override
     public long getTimeoutMilliseconds() {
-        // Should run quickly; increased to large timeout due to occasonal slow CI downloads
+        // Should run quickly; increased to large timeout due to occasoinal slow CI downloads
```
Comment (treo) [inline platform-tests/src/test/java/org/eclipse/deeplearning4j/dl4jcore/datasets/iterator/DataSetIteratorTest.java]: Typo.

File: platform-tests/src/test/java/org/eclipse/deeplearning4j/frameworkimport/tensorflow/TFGraphTestAllHelper.java [hunk 4]
```diff
@@ -123,6 +123,7 @@ public ModelLoadResult apply(File file, String name) {
                 SameDiff result = tensorflowFrameworkImporter.runImport(file.getAbsolutePath(), dynamicVariables, suggestDynamicVariables);
                 return new ModelLoadResult(result, graphDef);
             }catch(Exception e) {
+                e.printStackTrace();
```
Comment (treo) [inline platform-tests/src/test/java/org/eclipse/deeplearning4j/frameworkimport/tensorflow/TFGraphTestAllHelper.java]: Wouldn't it be better to log this?

Or is this a debugging leftover?

File: platform-tests/src/test/java/org/eclipse/deeplearning4j/frameworkimport/tensorflow/TFGraphTestAllSameDiff.java [hunk 5]
```diff
@@ -50,16 +50,18 @@
 
     /**
      * NOTE: If this is empty or the tests names are wrong,
-     * all tests will trigger an assumeFalse(..) that indicates
+     * all tests will triggr an assumeFalse(..) that indicates
```
Comment (treo) [inline platform-tests/src/test/java/org/eclipse/deeplearning4j/frameworkimport/tensorflow/TFGraphTestAllSameDiff.java]: typo

**Review by treo — CHANGES_REQUESTED** [review]

File: libnd4j/include/helpers/shape.h [hunk 6]
```diff
@@ -2016,9 +2016,15 @@ SD_INLINE SD_HOST_DEVICE sd::LongType strideAt(const sd::LongType *shapeInfo, co
  * @return
  */
 SD_INLINE SD_HOST_DEVICE bool equalsSoft(const sd::LongType *shapeA, const sd::LongType *shapeB) {
-  if (shapeA[0] != shapeB[0]) return false;
+  if (shapeA[0] != shapeB[0]) {
+    sd_printf("First case shape failure\n",0);
```
Comment (quickwritereader) [inline libnd4j/include/helpers/shape.h]: replace it with sd_debug if it needs to be logged

File: libnd4j/include/loops/impl/type_conversions.cpp [hunk 7]
```diff
@@ -1,20 +1,20 @@
...
+*  See the NOTICE file distributed with this work for additional
+*  information regarding copyright ownership.
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+* WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+* License for the specific language governing permissions and limitations
+* under the License.
+*
+* SPDX-License-Identifier: Apache-2.0
+******************************************************************************/
```
Comment (quickwritereader) [inline libnd4j/include/loops/impl/type_conversions.cpp:17]: if there is not any formatting benefits it could be unstaged

File: libnd4j/include/system/op_boilerplate.h [hunk 8]
```diff
@@ -1,20 +1,20 @@
...
+*  See the NOTICE file distributed with this work for additional
+*  information regarding copyright ownership.
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+* WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+* License for the specific language governing permissions and limitations
+* under the License.
+*
+* SPDX-License-Identifier: Apache-2.0
+******************************************************************************/
```
Comment (quickwritereader) [inline libnd4j/include/system/op_boilerplate.h:18]: could be unstaged if there is not any formatting

File: libnd4j/include/ops/declarable/impl/DeclarableOp.cpp [hunk 9]
```diff
@@ -335,7 +347,7 @@ int sd::ops::DeclarableOp::prepareOutputs(Context &ctx) {
           int shapeEquals = shape::equalsSoft(out, array->shapeInfo());
           int arrayEmpty = array->isEmpty();
           // checking out shape equality
-          if (!shapeEquals || arrayEmpty) {
+          if (!shapeEquals) {
```
Comment (quickwritereader) [inline libnd4j/include/ops/declarable/impl/DeclarableOp.cpp:350]: was ommiting "arrayEmpty check" here deliberate?

File: platform-tests/src/test/java/org/eclipse/deeplearning4j/frameworkimport/tensorflow/TFGraphTestAllHelper.java (hunk 4 above)
Comment (agibsonccc) [reply platform-tests/src/test/java/org/eclipse/deeplearning4j/frameworkimport/tensorflow/TFGraphTestAllHelper.java]: Actually no. This actually causes a silent failure to happen if I don't do that. Printing the stack trace during the test helps uncover why a graph couldn't be imported.

File: libnd4j/include/ops/declarable/impl/DeclarableOp.cpp (hunk 9 above)
Comment (agibsonccc) [reply libnd4j/include/ops/declarable/impl/DeclarableOp.cpp:350]: Yes. It was causing an error to be thrown when one didn't need to be. This came up in quite a few of the TF import tests. Shapes would be equal but then it would complain about something being invalid. Just because it's empty doesn't mean it's invalid.

File: platform-tests/src/test/java/org/eclipse/deeplearning4j/frameworkimport/tensorflow/TFGraphTestAllHelper.java (hunk 4 above)
Comment (treo) [reply platform-tests/src/test/java/org/eclipse/deeplearning4j/frameworkimport/tensorflow/TFGraphTestAllHelper.java]: Why not make it a loud error instead of printing to the console?

Comment (agibsonccc) [reply platform-tests/src/test/java/org/eclipse/deeplearning4j/frameworkimport/tensorflow/TFGraphTestAllHelper.java]: Done.

**Review by treo — APPROVED** [review]
